tests/unit/test_charm.py

```diff
@@ -269,3 +214,47 @@ def test__on_loki_push_api_endpoint_departed(self):
     #     self.assertEqual(
     #         self.harness.charm.unit.status, BlockedStatus("could not reload configuration")
     #     )
+
+    def test_loki_config_with_without_loki_endpoints(self):
+        self.harness.set_can_connect("agent", True)
+        rel_id = self.harness.add_relation("logging-consumer", "loki")
+
+        self.harness.add_relation_unit(rel_id, "loki/0")
+        endpoint0 = json.dumps({"url": "http://loki0:3100:/loki/api/v1/push"})
+        self.harness.update_relation_data(rel_id, "loki/0", {"endpoint": endpoint0})
+
+        self.harness.add_relation_unit(rel_id, "loki/1")
+        endpoint1 = json.dumps({"url": "http://loki1:3100:/loki/api/v1/push"})
+        self.harness.update_relation_data(rel_id, "loki/1", {"endpoint": endpoint1})
+
+        expected = {
+            "loki": {
+                "configs": [
+                    {
+                        "name": "promtail",
+                        "clients": [
+                            {"url": "http://loki0:3100:/loki/api/v1/push"},
```

Comment:
`loki0:3100:` seems invalid, which makes me wonder if url validation should take place in the charm/lib.

Reply:
In this test we are checking that the URLs sent by the Provider side of the relation are obtained by the Consumer side...

I think there is no need to validate URL since we are using `socket.getfqdn()`:

```python
    @property
    def _url(self) -> str:
        """Get local Loki Push API url.

        Return url to loki, including port number, but without the endpoint subpath.
        """
        return "http://{}:{}".format(socket.getfqdn(), self.port)
```